storage/registry.go

```diff
@@ -18,66 +18,57 @@ package storage
 
 import (
 	"context"
-	"os"
-	"os/signal"
 	"sync"
-	"syscall"
 
-	"github.com/Sirupsen/logrus"
+	"fmt"
+
+	"github.com/sirupsen/logrus"
 )
 
 var (
 	mux      sync.RWMutex
 	storages = make(map[string]Storage)
 )
 
-// Get returns a single storage
-// Panics if more than one storage is configured. Use GetByName in such cases
-func Get() Storage {
+func Register(name string, storage Storage) {
 	mux.RLock()
 	defer mux.RUnlock()
-	storagesCount := len(storages)
-	if storagesCount != 1 {
-		logrus.Panicf("Requested exactly one storage but %d storages are configured", storagesCount)
+	if storage == nil {
+		panic("storage: Register storage is nil")
 	}
-	var registeredStorage Storage
-	for _, v := range storages {
-		registeredStorage = v
-		break
+	if _, dup := storages[name]; dup {
+		panic("storage: Register called twice for storage with name " + name)
 	}
-	return registeredStorage
-}
-
-// GetByName returns a storage with this name and boolean indicating whether it exists
-func GetByName(name string) (Storage, bool) {
-	mux.RLock()
-	defer mux.RUnlock()
-	providedStorage, exists := storages[name]
-	return providedStorage, exists
+	storages[name] = storage
 }
 
-// Use specifies the storage for the given name
-func Use(name string, storage Storage, ctx context.Context) {
+// Argumentation for refactoring
+//Use returning the storage seems more natural rather then doing .Get; .Get() introduces a hidden dependency as well)
+//This way we kind of provide a wrapper over the sqlx.db by keeping the same flow of usage while keeping an abstraction for Storage possible
+// Wrapping as follows
+// - postgres impl of storage _ imports the driver, we _import the postgres storage's package
+// Registering a storage "wraps" registering a driver and allows us to "use" the storage that wraps abstract opening a storage (in posgres it opens sqlx.db)
+// The idea is to allow abstraction of Storage while keeping the sqlx.db flow of usage as close as possible
+// uri is passed as argument as we don't want to obtain env variables in random places of the code as the config might not be provided as env vars)
+// It's better to abstract away the environment loading and use it to build a configuration that is passed to the server. The server than takes care
+// to split the configuration and pass segments to the components it initializes (to storage the server will pass uri)
+func Use(name string, uri string, ctx context.Context) (Storage, error) {
```

Comment:
I'm ok with this way. It would work with postgres, mongo, etc. but etcd can take more than 1 uri for different nodes
If we go the db/sql way we should register something that would provide you a Storage. Then on Use(..) you'd have storageProvider.Provide(uri?)  which would give you the new Storage. Not sure if we'd benefit from separating the opening from the closing of the storage

Reply:
Right now
We regisrter a Storage that during registration is a "wrapper" for driver registration
Then we Use the Storage which during "usage/opening" is wrapper for sql/db opening

So we wrap both the "driver" and the "db" with our storage which seems fine to me - 

if we see a benefit we may split it so that we "wrap" the driver with storage and the "db.opening" with provider - but right now I can't think of a decent use case?